### ISSUE TITLE
Option to run tests in parallel

### ISSUE THREAD

**Issue by AndreyBelym** [DevExpress/testcafe#1165]
### Are you requesting a feature or reporting a bug?
Feature

### What is the current behavior?
TestCafe can run test only in series.

### What is the expected behavior?
TestCafe can run tests concurrently.

**Comment by AndreyBelym** [comment]
## Feature Proposal
### Interface
Inspired by [AVA](https://github.com/avajs/ava)

  - Command line key `-c/--concurrent <concurencyNumber>` + API Runner method `.concurrent(<concurencyNumber>)`

 - `.serial` property for tests and fixtures to disallow running them concurrently

### Implementation
  - Should start `<numberOfBrowsers>*<concurencyNumber>` browsers. *Remotes are problematic part here*.

  - Current architecture allows different BrowserConnections (send the test URL to the browser) to share the same BrowserJob (generates test URLs). The only problem here is to filter `.serial` tests.

**Comment by inikulin** [comment]
As far as I remember reporter relies on serial execution as well.

**Comment by inikulin** [comment]
BTW, why do we need `serial` in the first place? What's the use case? Pure serial execution was never guaranteed by TestCase (same test runs in parallel in different browsers) and it worked well so far, so what itch we try to scratch here? Moreover, it will not work well with our reporting: since we have text-based reporters we can't break the order of tests. If you run "serial" tests first: you either don't report them immediately and wait for their fixture to start, or somehow split report in two parts: one for serial and one for parallel tests.

**Comment by AlexanderMoskovkin** [comment]
@AndreyBelym have updated the post, now it's 

> - `.serial` property for tests and fixtures to disallow running them concurrently. They should not change overall test execution order, just prevent next tests to be started.
```
test('1', () => );
test('2', () => ...);
test.serial('3', () => );
test('4', () => ...);
test('5', () => ...);
test.serial('6', () => );

runner.concurrent(2).run() =>

test1  test2
test3
test4  test5 
test6
```
It seems this approach doesn't break reporting.
In theory the `concurrently` feature can be useful when you run tests in a single browser and some tests can be run concurrently but some tests can't be (for example they change a database state). In this case you can mark several tests as `serial` and run others concurrently.

BUT, this option has some uncertainties.
For example if I run tests just in two browsers: `testcafe chrome,firefox test.js`, and I have a `serial` test. Should it be run in two browsers concurrently or it should be run one by one? How it should work if I run tests in two browsers with the concurrently option via `testcafe chrome,firefox test.js -c 2` ?
So, do we really need the `single` test feature? /cc @DevExpress/testcafe

**Comment by inikulin** [comment]
> In theory the concurrently feature can be useful when you run tests in a single browser and some tests can be run concurrently but some tests can't be (for example they change a database state). 

Please, read carefully: we already have this behavior. Don't we try to solve a problem that doesn't exists here?

> For example if I run tests just in two browsers: testcafe chrome,firefox test.js, and I have a serial test. Should it be run in two browsers concurrently or it should be run one by one? How it should work if I run tests in two browsers with the concurrently option via testcafe chrome,firefox test.js -c 2 ?

I have a great solution for this problem: don't add `serial`.

**Comment by guptaakhil** [comment]
Please fo let us know the exact syntax to run or share sample code , very much appreciate if some one can.

**Comment by miherlosev** [comment]
See https://devexpress.github.io/testcafe/documentation/using-testcafe/common-concepts/concurrent-test-execution.html

**Comment by guptaakhil** [comment]
It's not working for us, Tried severa time only serial execution is performing. I beleive there is some workaround to acheive this as they are opening in same port that could be run they are executing in serially.

**Comment by miherlosev** [comment]
@guptaakhil 
It is difficult to understand the reason of the problem by provided your information .
Please, create a separate issue and specify operation system name, nodejs and testcafe versions.
In issue description telling us how you run the tests.

**Comment by guptaakhil** [comment]
sure, we will be writing this into seperate thread.
